articles/actions.md

```diff
@@ -0,0 +1,338 @@
+A user-define description of the goal is sent as the request.
+The response indicates whether or not the goal was accepted, and if so the identifier the server will use to describe the goal.
+
+The QoS settings of this service should be set the so the client is guaranteed to receive a response or an action could be executed without a client being aware of it.
+
+### Cancel Request Service
+
+* **Direction**: Client calls Server
+* **Request**: Goal identifier, time stamp
+* **Response**: Goals that will be attempted to be canceled
+
+The purpose of this service is to request to cancel one or more goals on the action server.
+A cancellation request may cancel multiple goals.
+The result indicates which goals will be attempted to be canceled.
+Whether or not a goal is actually canceled is indicated by the status topic and the result service.
+
+The cancel request policy is the same as in ROS 1.
+
+* If the goal ID is empty and time is zero, cancel all goals
+* If the goal ID is empty and time is not zero, cancel all goals accepted at or before the time stamp
+* If the goal ID is not empty and time is not zero, cancel the goal with the given id regardless of the time it was accepted
+* If the goal ID is not empty and time is zero, cancel the goal with the given id and all goals accepted at or before the time stamp
+
+### Get Result Service
+
+* **Direction**: Client calls Server
+* **Request**: Goal ID
+* **Response**: Status of goal and user defined result
+
+The purpose of this service is to get the final result of a service.
+After a goal has been accepted the client should call this service to receive the result.
+The result will indicate the final status of the goal and any user defined data.
+
+Once the server sends the result to the client it should free up any resources used by the action.
+If the client never asks for the result then the server should discard the result after a timeout period.
+
+### Goal Status Topic
+
+* **Direction**: Server publishes
+* **Content**: List of in-progress goals with: Goal ID, time accepted, and an enum indicating the status
+
+This topic is published by the server to broadcast the status of goals it has accepted.
+The purpose of the topic is for introspection; it is not used by the action client.
+Messages are published when transitions from one status to another occur.
+
+The default QoS settings for a DDS middleware should be TRANSIENT LOCAL with a history size of 1.
+This allows new subscribers to always get the latest state.
+
+The possible statuses are:
+
+* *Accepted*
+  * The goal has been accepted by the action server and may now be executing
+* *Cancelling*
+  * The action server will try to cancel the indicated goal
+* *Cancelled*
+  * The action server successfully canceled the goal
+* *Succeeded*
+  * The action server successfully reached the goal
+* *Aborted*
+  * The action server failed reached the goal
+
+### Feedback Topic
```

Comment:
It looks like this is mirroring the concept of a single 'feedback' topic from ROS1. This is really suboptimal in terms of bandwidth and processing. If you have one busy action server sending feedback to many clients, the server ends up sending out data about M goals to N clients.

Is there any reasonable mechanism that would allow the clients to only receive/process the feedback they care about?

Reply:
The only thing that comes to mind is using topic namespaces to reduce overhead. For example, we could use some unique information about the client as part of the feedback/status topic names:

```
_action/fibonacci/<client_id>/feedback
_action/fibonacci/<client_id>/status
```

Any other thoughts/suggestions on the matter are welcome.

Reply:
It's possible this solution increases overhead in the middleware, but reduces bandwidth and processing for the nodes?

Reply:
You could specify the feedback/status endpoint in the goal I suppose. Per-process endpoints (e.g. client_id, or node_name) could be the right balance between bandwidth overhead and middleware overhead.

What is middleware overhead in this context? The number of topics?

Reply:
Allowing the client to specify a custom feedback topic would be useful to alleviate the extreme cases I pointed out, but there's also value in being able to know what the feedback topic will be without fail (for things like recording). It would also complicate the ROS interface between action client and server, as well as complicating the implementation of the action server.

That being said it might be worth doing.

> What is middleware overhead in this context? The number of topics?

Yes, but I actually think of it as two dimensional. First, the number of topics obviously increases the memory overhead and potentially number of sockets used. Second, the creation and destruction of new topics has overhead (think of it like threads). This might be alleviated with "topic pools" which are reused overtime, but that would not be complicated if the client choose the topic.

Basically, every time you create a pub or sub on a topic there is discovery information and QoS/type matching that needs to occur which is asynchronous and relatively expensive (sometimes involving broadcasting information to whole graph, also requiring the creation of new sockets, etc...). It would also likely increase the latency between "goal accepted" and first "goal feedback received", assuming feedback would be immediately available.

Reply:
I'm not keen on allowing clients to specify the feedback topic as a solution to the feedback fire hose problem. I think this requires a design decision in the implementation of the client that may not be decidable when that client is implemented.

To be specific, let's say I have an existing action server I want to use and it is popular in my system - its actions are called frequently from many clients. Furthermore, some of those clients are developed by me and some are not. I can make my clients declare a personal feedback topic to avoid the information overload, but I cannot guarantee that the other clients do. Their developers may have assumed that the action would be used only by one or a couple of clients at a time, for example. This may mean that those other clients get overloaded with probably irrelevant feedback because the assumption made when they were developed is no longer valid.

So to summarise, if clients specifying the feedback topic is how the feedback volume problem will be solved, then it needs to be a requirement for all clients, not an option.

I am not opposed to allowing clients to specify the feedback topic in general, just as a solution to this particular problem.

Reply:
It looks like IDL is landing in Crystal. Would it be possible to have it used in the action implementation off the bat?

Reply:
I've captured this conversation in the "alternatives" section of the document.